src/main/java/com/aws/greengrass/config/ConfigurationReader.java

```diff
@@ -103,6 +103,38 @@ private static void mergeTLogInto(Configuration c, Path p) throws IOException {
         }
     }
 
+    /**
+     * In place update for the given config object from the given transaction log, adhering to the given update behavior
+     * tree and without losing listeners.
+     *
+     * @param config         configuration to merge into
+     * @param tlogPath       path of the tlog file to read to-be-replaced config from
+     * @param forceTimestamp should ignore if the proposed timestamp is older than current
+     * @param mergeCondition Predicate that returns true if the provided Topic should be merged and false if not
+     * @param tree           Merge behavior hierarchy for the update
+     * @throws IOException if update fails
+     */
+    public static void updateFromTLog(Configuration config, Path tlogPath, boolean forceTimestamp,
+                                      Predicate<Node> mergeCondition, UpdateBehaviorTree tree) throws IOException {
+
+        // Merge tlog into configuration so that nodes to retain get replaced
+        ConfigurationReader.mergeTLogInto(config, tlogPath, forceTimestamp, mergeCondition);
+
+        // Remove extraneous nodes from configuration
+        ConfigurationReader.discardNodesNotInTLog(config, tlogPath, tree);
+    }
+
+    private static void discardNodesNotInTLog(Configuration config, Path tlogPath, UpdateBehaviorTree tree)
+            throws IOException {
+        Configuration tlogMirror = createFromTLog(new Context(), tlogPath);
+
+        config.deepForEach((n, b) -> {
+            if (UpdateBehaviorTree.UpdateBehavior.REPLACE.equals(b) && tlogMirror.findNode(n.path()) == null) {
```

Comment:
technically you should be looking up the behavior for this node in particular right?

Reply:
`b'` is the node's parent's behavior, and parent's behavior determines if the node should be removed so the parent gets 'replaced' as a result.

Reply:
But I don't know how it will move past the root node this way, I'll know when I test it

Reply:
I take this back, it needs to be the node, not its parent